### PR TITLE
Fix additional_table_filters for Distributed.

### LINKED ISSUE #41692: additional_table_filters with distributed table
This impacts the Grafana plugin.

From https://github.com/grafana/clickhouse-datasource/issues/198

    if an adhoc filter is specified for the field t1_distributed.f2
        the following query will be built by the plugin SELECT f1 from t1_distributed settings additional_table_filters = {'t1_distributed': 'f2 = xxx'}
        this query will actually not do its job of filtering
        i guess it is coming from the fact that additional_table_filters are applied by clickhouse when it reads the table....but distributed tables are not read.....
        if i replace in the built query, within the additional_table_filters settings, t1_distributed with t1_local (so a query like `SELECT f1 from t1_distributed settings additional_table_filters = {'t1_local': 'f2 = xxx'}), then it is working as expected
    i gave a try to add an adhoc filter on the "local" table (while my "main" query is still on the distributed table) (ie define an adhoc filter on the fied t1_local.f2)
        then the built query by the plugin does not include any additional_table_filters settings (i guess there is a check in order to add additional_table_filters only on the table used in the query)

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Setting `additional_table_filters` were not applied to `Distributed` storage. Fixes #41692

### CODE REVIEW

File: src/Interpreters/ClusterProxy/executeQuery.cpp [hunk 1]
```diff
@@ -96,6 +97,14 @@ ContextMutablePtr updateSettingsForCluster(const Cluster & cluster, ContextPtr c
         new_settings.limit.changed = false;
     }
 
+    if (query_info && query_info->additional_filter_ast)
```
Comment (alexey-milovidov) [inline src/Interpreters/ClusterProxy/executeQuery.cpp:100]: Missing comments.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: src/Interpreters/ClusterProxy/executeQuery.cpp [hunk 2]
```diff
@@ -96,6 +97,14 @@ ContextMutablePtr updateSettingsForCluster(const Cluster & cluster, ContextPtr c
...
+    {
+        Tuple tuple;
+        tuple.push_back(main_table.getFullTableName());
+        tuple.push_back(queryToString(query_info->additional_filter_ast));
+        new_settings.additional_table_filters.value.push_back(std::move(tuple));
```
Comment (alexey-milovidov) [inline src/Interpreters/ClusterProxy/executeQuery.cpp:105]: Why it's not propagated automatically as every other query setting?

**Comment by alexey-milovidov** [comment]
Non obvious.

File: src/Interpreters/ClusterProxy/executeQuery.cpp (hunk 2 above)
Comment (KochetovNicolai) [reply src/Interpreters/ClusterProxy/executeQuery.cpp:105]: It is propagated. But we need to change table name for distributed.

**Comment by KochetovNicolai** [comment]
Ast fuzzer:
```
create table table_1__fuzz_37 (x Nullable(UInt128)) engine = MergeTree order by x settings allow_nullable_key=1;
insert into table_1__fuzz_37 values (1), (3);
SELECT x + 1 FROM table_1__fuzz_37 SETTINGS additional_result_filter = '`plus(x, 1)` != 2';

Received exception from server (version 22.10.1):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Arguments of 'plus' have incorrect data types: '' of type 'Nullable(UInt128)', '' of type 'UInt8'. (LOGICAL_ERROR)

```

File: src/Interpreters/ClusterProxy/executeQuery.cpp [hunk 3]
```diff
@@ -96,6 +97,20 @@ ContextMutablePtr updateSettingsForCluster(const Cluster & cluster, ContextPtr c
...
+    /// Setting additional_table_filters may be applied to Distributed table.
+    /// In case if query is executed up to WithMergableState on remote shard, it is impossible to filter on initiator.
+    /// We need to propagate the setting, but change the table name from distributed to source.
+    ///
+    /// Here we don't try to analyze setting again. In case if query_info->additional_filter_ast is not empty, some filter was applied.
+    /// It's just easier to add this filter for a source table.
+    if (query_info && query_info->additional_filter_ast)
+    {
+        Tuple tuple;
+        tuple.push_back(main_table.getFullTableName());
```
Comment (alexey-milovidov) [inline src/Interpreters/ClusterProxy/executeQuery.cpp:109]: Does it work correctly if the cluster is using different database names on different shards?

**Comment by alexey-milovidov** [comment]
Should it be fixed for Merge and Buffer tables?
Does it work for Views?

File: src/Interpreters/ClusterProxy/executeQuery.cpp (hunk 3 above)
Comment (KochetovNicolai) [reply src/Interpreters/ClusterProxy/executeQuery.cpp]: Added a test.

**Comment by KochetovNicolai** [comment]
Merge, View and Mat. View already were covered in test.
Buffer table does not proxy a query, only reading part. Filter is added after.

**Comment by KochetovNicolai** [comment]
AST Fuzzer case will be fixed in #42410

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**